specs/WinRT/MakeMSIX-spec.md

```diff
@@ -0,0 +1,959 @@
+        INT64 versionPartsRequired = 4;
+        INT64 versionDigitsToAdd = versionPartsRequired - versionPartsFound;
+        for (auto i = 0; i < versionDigitsToAdd; i++)
+        {
+            productVersionString += L".0";
+        }
+        std::wstring processorArchitecture{};
+        SYSTEM_INFO systemInfo;
+        GetNativeSystemInfo(&systemInfo);
+        switch (systemInfo.wProcessorArchitecture)
+        {
+        case PROCESSOR_ARCHITECTURE_AMD64:
+            processorArchitecture += L"x64";
+            break;
+        case PROCESSOR_ARCHITECTURE_ARM:
+            processorArchitecture += L"arm";
+            break;
+        case PROCESSOR_ARCHITECTURE_ARM64:
+            processorArchitecture += L"arm64";
+            break;
+        case PROCESSOR_ARCHITECTURE_INTEL:
+            processorArchitecture += L"x86";
+            break;
+        case PROCESSOR_ARCHITECTURE_IA64:
+        case PROCESSOR_ARCHITECTURE_UNKNOWN:
+            winrt::throw_hresult(E_UNEXPECTED);
+        }
+
+        std::filesystem::path platformSDKExecutablePath = installPathString;
+        platformSDKExecutablePath /= L"bin";
+        platformSDKExecutablePath /= productVersionString;
+        platformSDKExecutablePath /= processorArchitecture;
+
+        return platformSDKExecutablePath;
+    }
+
+    void LaunchSignToolWithArguments(std::wstring arguments)
+    {
+        std::filesystem::path signToolPath = GetPlatformSDKPath();
+        signToolPath /= L"signtool.exe";
+
+        std::wstring commandLine{ signToolPath };
+        commandLine += L" " + arguments;
+        DWORD exitCode{};
+        winrt::check_hresult(CreateProcessAndWaitForExitCode(commandLine, exitCode));
+        if (exitCode == EXIT_FAILURE)
+        {
+            winrt::throw_hresult(E_FAIL);
+        }
+    }
+
+    void Example_SignPackage(std::wstring pfxFile, std::wstring packageFile)
+    {
+        std::wstring signToolArguments;
+        signToolArguments.append(L"sign /fd SHA256 /f ");
+        signToolArguments.append(pfxFile);
+        signToolArguments.append(L" " + packageFile);
+        LaunchSignToolWithArguments(signToolArguments);
+    }
+```
+
+# 5. API Details
+
+```c#
+// Copyright (c) Microsoft Corporation and Contributors.
+// Licensed under the MIT License.
+
+namespace Microsoft.Kozani.MakeMSIX
+{
+    [contractversion(1)]
+    apicontract MakeMSIXContract{};
+
+    /// Options for creating a package
+    [contract(MakeMSIXContract, 1)]
+    runtimeclass PackOptions
+    {
+        /// <summary>
+        /// Created by callers to pass options into CreatePackage
+        /// </summary>
+        PackOptions();
+
+        /// <summary>
+        /// Output path for the full packaged file.
+        /// </summary>
+        String PackageFilePath{ get; set; };
+
+        /// <summary>
+        /// Overwrite PackageFilePath if it already exists.
+        /// Defaults to true.
+        /// </summary>
+        Boolean OverwriteFiles{ get; set; };
```

Comment:
I understand how Prompt would work in the context of makeappx or any command line tool, but not sure how it would work for a winrt API. It seems reasonable to let the caller check the filesystem for existence of the file first if they care. Barring that, the api prototype currently checks for existence of the output file up front before doing any work and returns HRESULT(ERROR_FILE_EXISTS) when the file exists and Overwrite is false, which the caller could use to show a prompt.

Reply:
Are there other apis in WindowsAppSDK that pop UI up? I haven't looked at what it would take to localize ui and make that work in the WindowsAppSDK context.

Reply:
Ahhh. My bad. I mixed up the API and the wrapper tool.

Prompting is entirely a tool thing. The API is a boolean. Any prompting should be handled by the tool e.g.

MAKEMSIX PACK --Overwrite=Prompt... == IF file exists Prompt for Y/N then call API with options.OverwriteFileIfExists=Y/N